[Feature] Cart CRUD 구현

growingpsy opened the pull request:
## 🛠️ Issue Number
### closes #85 

📌 **작업 내용 및 특이사항**

**Cart Create**
장바구니 추가 (POST /carts)

* 회원 ID, 옵션 값 ID, 수량으로 신규 장바구니 생성

* 동일 옵션 상품이 이미 존재할 경우 수량만 합산

* Cart.createCart() 정적 메서드로 엔티티 생성

* 권한: CUSTOMER

**Cart Read**
장바구니 전체 조회 (GET /carts?memberId=)

* 회원별 전체 장바구니 목록 조회

* CartResponse.from()으로 응답 변환

**Cart Update**
장바구니 수량 수정 (PATCH /carts/{id})

* 수량 동일 시 CART_NOT_MODIFIED 예외 발생

* 권한: CUSTOMER

**Cart Delete**
장바구니 단일 삭제 (DELETE /carts/{id})

* 권한: CUSTOMER

- Cart Item 통합 조회 (상품 정보, 옵션명, 이미지 포함)

📚 **참고사항**

growingpsy committed:
refactor #85: Cart 엔티티에 ProductOptionValue 매핑

growingpsy committed:
chore #85: spotless apply, 불필요 파일 정리

growingpsy committed:
feature #85: Cart CRUD 구현

growingpsy committed:
refactor #85: CartService에서 ProductOptionValueRepository 적용 및 검증 로직 수정

- ProductRepository 의존 제거
- ProductOptionValueRepository 주입 및 findById()로 옵션값 검증
- Cart 생성 시 옵션값 유효성 검사 로직 개선

haeda0225 committed:
[Feature] Category crud 개발 (#145)

* feature #116: category crud 개발

* feature #116: category crud 개발_1차 pr 리뷰 수정반영

hyeonji91 committed:
feature #170 : 상점-배송비정책 외래키를 상점이 보유

hyeonji91 committed:
style #170 : spotless 적용

growingpsy committed:
refactor #85: CartResponse에 가격 계산 필드 추가

growingpsy committed:
refactor #85: 장바구니 조회 시 상품 및 옵션 정보 포함하도록 쿼리 개선

- findAllWithProductByMemberId 쿼리 추가 (상품, 옵션 그룹 정보 fetch join)
- 기존 단순 목록 조회(findAllByMemberId) 대체
- 추후 상품 이미지 조인(LEFT JOIN FETCH p.productImages) 추가 예정

growingpsy committed:
refactor #85: 장바구니 수량 합산 로직 및 조회 쿼리 개선

- 기존 createCart 로직 개선: 기존 수량(oldQuantity)과 신규 수량 합산 처리
- 로깅 정보 세분화(memberId, optionValueId, oldQuantity, newQuantity)
- getCartListByMember에서 상품/옵션 정보 포함 조회(findAllWithProductByMemberId)로 변경

willjsw commented on src/main/java/com/irum/come2us/domain/cart/application/service/CartService.java:
```diff
@@ -0,0 +1,101 @@
+        List<Cart> carts = cartRepository.findAllByMemberId(memberId);
+        log.info("장바구니 조회 완료: memberId={}, count={}", memberId, carts.size());
+        return carts.stream().map(CartResponse::from).collect(Collectors.toList());
+    }
```
P2: 해당 로직에서 cart 조회 시 1번, stream 메서드 참조를 통한 ResponseDTO 생성 과정에서 추가 쿼리 발생으로 cart.size() 만큼의 조회 쿼리 N번이 추가로 발생하는 N+1 문제가 있습니다. findAllByMember 로직에 fetchJoin 적용하거나, DTO에서 .memberId(cart.getMember().getMemberId()) 을 통해 업데이트 하는 Member 필드를 반환값에서 삭제하는 것도 검토해볼 만 한 것 같습니다.

growingpsy replied:
CartResponse에서 memberId 필드를 제거하고, findAllWithProductByMemberId()에 fetch join을 적용하여 N+1 쿼리 문제가 발생하지 않도록 수정했습니다. 현재 조회 로직은 JOIN FETCH로 Product, OptionGroup, OptionValue까지 한 번에 조회되며, Member는 로그인 사용자 기준(MemberUtil)으로 고정되어 추가 조회 없이 단일 쿼리로 장바구니 전체 데이터를 가져오도록 개선되었습니다👍

willjsw commented on src/main/java/com/irum/come2us/domain/cart/application/service/CartService.java:
```diff
@@ -0,0 +1,101 @@
+        List<Cart> carts = cartRepository.findAllByMemberId(memberId);
+        log.info("장바구니 조회 완료: memberId={}, count={}", memberId, carts.size());
+        return carts.stream().map(CartResponse::from).collect(Collectors.toList());
+    }
+
+    public void deleteCart(UUID cartId) {
+        Cart cart =
+                cartRepository
+                        .findById(cartId)
+                        .orElseThrow(() -> new CommonException(CartErrorCode.CART_NOT_FOUND));
+
+        cartRepository.delete(cart);
+        log.info("장바구니 삭제 완료: cartId={}", cartId);
+    }
+}
```
P1: MemberUtil 통한 사용자 리소스 접근 권한 확인 로직 추가 부탁드립니다! delete 외 update 쪽에도 동일한 적용 필요합니다.

willjsw commented on src/main/java/com/irum/come2us/domain/cart/application/service/CartService.java:
```diff
@@ -0,0 +1,101 @@
+        Member member =
+                memberRepository
+                        .findByMemberId(request.memberId())
+                        .orElseThrow(() -> new CommonException(CartErrorCode.MEMBER_NOT_FOUND));
```
P1: MemberId를 request를 통해 받게 되면 인증 필터가 무의미해집니다. MemberUtil의 assertMemberResourceAccess 메서드 통해 처리해주시길 바랍니다! 
#157 참고바랍니다:)

growingpsy replied:
MemberUtil 반영했으니 확인부탁드립니다!

willjsw reviewed (changes requested):
수정사항 확인 부탁드려요! 조금만 더 힘냅시다👊

growingpsy committed:
style #85: spotless apply

growingpsy committed:
refactor #85: CartService에 로그인 사용자 권한 검증 로직 추가 및 CartErrorCode 확장

growingpsy committed:
style #85: spotless apply

growingpsy committed:
refactor #85: MemberUtil 기반 본인 장바구니 전용 구조로 리팩토링

- CartService에 MemberUtil 적용하여 로그인 사용자 기준으로 동작하도록 수정
- memberId 파라미터 제거 및 불필요한 MemberRepository 의존성 제거
- CartResponse에서 memberId 필드 삭제 (N+1 방지 및 보안 강화)

willjsw reviewed (approved):
고생하셨습니다:)
